### ISSUE TITLE
Eigenvector centralization is problematic

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2679]
In a previous issue, I discussed one reason why eigenvector centralization is problematic: #2673 (it depends on the normalization of the vertex-level centralities).

There are other conceptual problems as well.

The most centralized structure is quoted in the documentation as:

 - Undirected: A graph with a single edge (for any number of vertices).
 - Directed: The in-star.

Neither of these are connected graphs. Eigenvector centrality is not suitable for disconnected graphs: the solution is degenerate. Then does it make sense to define centralization based on a disconnected structure?

In the directed case in particular, igraph returns all-zeros for any directed acyclic graph, _including_ the "most centralized" in-star. In cases of DAGs, the only eigenvalue is 0, and the solution is degenerate. Returning all-zeros is not unreasonable, but it is not consistent with calling the in-start the most centralized.

Generally, the concept of eigenvector centralization seems extremely dubious, and badly thought out. Despite this, there are even recent papers written on it (e.g. [this](https://doi.org/10.1080/0022250X.2021.1878357)), though these seem a bit naive to me, and don't try to be mathematically principled. 

I would prefer to remove this measure from igraph. However, people still use this measure frequently for social network analysis. That can be taken as a reason for, or against removal, depending on your perspective.

What are your thoughts on this topic?

CC @vtraag @mbojan

**Comment by szhorvat** [comment]
Regarding DAGs: Consider vectors where all elements are zero except at a single sink vertex. Any linear combination of these is a valid solution.

The approach we were moving towards for 1.0 was not to fuss too much with non-(strongly-)connected graphs: warn the user that eigenvector centrality is unsuitable for this graph, and return _any_ eigenvector paired to the leading eigenvalue.

**Comment by szhorvat** [comment]
I made a change in the computation of eigenvector centralities for directed acyclic graphs. These were special-cased, likely because ARPACK may choke on them (though the reason was not documented). Previously, igraph would return an all-zero vector here. Now it returns a vector that has 1s in sink vertices and 0s everywhere else. This is one valid eigenvector, and is consistent with treating the in-start as the most centralized directed network when computing the theoretical maximum of the eigenvector centralization.

I still maintain that eigenvector centralization is a very poor concept (that likely sees much misuse), this change at least improves how igraph handles it ...

See the above commit.

Also CC @schochastics for input as you are more experienced with both social network analysis and centralities in general.

**Comment by schochastics** [comment]
I can only speak confidently about the undirected case, because that's where most of my work is situated.

I dont think it makes much sense that the graph with one edge and n-2 isolates should be the most centralized. Given that the concept of eigenvector centrality is not well defined for disconnected graph, the very least should be a warning when eigen centralization is used on a disconnected graphs, or even an error. 

In terms of normalization: is there a specific example where this is problematic? What I am thinking is that if there exist two graphs, `g1` and `g2` where depending on the normalization either `g1` or `g2` is more centralized, then it is indeed problematic. If only the magnitude of the difference changes then it would not be an issue. I am an advocate of centrality (and centralization) being on an ordinal scale, not interval. But I think I was not very successful in spreading that :wink:

**Comment by szhorvat** [comment]
> the very least should be a warning when eigen centralization is used on a disconnected graphs, or even an error

I fully agree, and this has been changed already for igraph 1.0 (not yet released). Some discussion is in PR #2592

But I do have to say that the situation can get a bit more complicated, and I think it does warrant more careful consideration ... I'm interested in your opinion on the following, @schochastics. 

Hub and authority scores are essentially the eigenvector centralities of the bibliographic coupling and cocitation networks. It is not that easy to say when these (undirected) networks will be connected by just looking at the original directed network they are derived from. In fact, it very often won't be connected, even if the original graph was strongly connected. They won't be connected in many real-world scenarios.

More interestingly, the largest connected component of the cocitation and bibcoupling networks may not even be the same. This happens despite the fact that the two scores are coupled (one can be computed from the other, and the associated eigenvalue is the same). Example:

<img width="360" alt="image" src="https://github.com/user-attachments/assets/b4f40b62-1a1c-4c68-8908-a3272a246306">

Cocitation:

<img width="360" alt="image" src="https://github.com/user-attachments/assets/859393a4-403a-46f7-9262-206e080e6e32">

Bibcoupling:

<img width="360" alt="image" src="https://github.com/user-attachments/assets/51c5867b-ba65-47f4-b740-4e9120dc9cdb">

What we try to do is to warn only if there isn't a "big enough" connected component of the bibcoupling / cocitation networks.

In more detail, the solution we went with for igraph 1.0 is the following, in brief:

 - We care only about non-negative weights (for now we warn about negative ones, but this may become an error in the future).
 - We define eigenvector centrality as a non-negative solution $x$ of $Ax = \lambda x$. Note the subtle difference from another common definition: the point is not to take the leading eigenvector, but to take a non-negative eigenvector (that's not all-zero). Sometimes, there's more than one of these. In fact, this turns out to happen precisely when the graph is not connected. Typically, the centrality scores for all but one component will be zeros (there's a potential exception when there are two components with exactly the same eigenvalue). Thus we can indirectly detect non-connected graphs by looking for zeros in the scores. This seems to work well for practical purposes.
 - When computing eigenvector centrality, we warn about disconnectedness even when there's a single zero score, and point out that the solution is non-unique.
 - When computing hub or authority scores, we are a bit more pragmatic and only warn when the fraction of zeros exceeds a certain limit. But do we consider this in the hub score or authority scores? Right now, only one of the two are tested. There are weird edge-cases, such as directed star graphs, where one of these two scores is almost all zeros, and the other is the opposite: almost all non-zeros.

**Comment by szhorvat** [comment]
> In terms of normalization: is there a specific example where this is problematic? What I am thinking is that if there exist two graphs, `g1` and `g2` where depending on the normalization either `g1` or `g2` is more centralized, then it is indeed problematic. If only the magnitude of the difference changes then it would not be an issue. I am an advocate of centrality (and centralization) being on an ordinal scale, not interval. But I think I was not very successful in spreading that 😉

This is a very good point, and I wasn't sure about the answer. Unfortunately, it turns out that even the ordering is not preserved.

Consider these two vectors: v1 = (3, 0, 4) and v2 = (5, 1, 3).

If we normalize by the maximum, and compute the sum of differences from the max score (i.e. centralization), we can the higher centralization for v1. If we normalize by the Euclidean norm, we get a higher centralization for v2.

So the concept really is messed up :-(

**Comment by schochastics** [comment]
> We define eigenvector centrality as a non-negative solution x of A x = λ x . 

Seems sensible. I think Martin Everett has some work on disconnected eigenvector centrality which was never published. I will reach out to him.

> So the concept really is messed up :-(

Going through the literature, I found more that make it worse. From: <https://doi.org/10.1016/S0378-8733(00)00031-9>:

``` r
library(igraph)
g <- make_star(11,"undirected")
h <- make_star(10, "undirected")
h <- add_vertices(h,1)
h <- add_edges(h,c(10,11))
centr_eigen(g)$centralization
#> [1] 0.7597469
centr_eigen(h)$centralization
#> [1] 0.7618274
```
So the star is not even the most centralized graph. From the paper:
> For the maximum norm, [...], this maximum is not known
yet but it can be seen from the example that the maximum is not reached by the star [...] 
Even worse, evaluation of some instances of n shows that
the structure determining the maximal value changes with n. That is for n = 4 it is the center
of the star, and for n = 11 it is definitely not the star

Towards the end of the paper they suggest to use the euclidean norm to normalize eigenvector centrality since this provides the desired properties. Apparently this is also mentioned in here: <https://doi.org/10.1016/S0378-8733(96)00301-2>

Maybe it needs some more experimenting but what I got from reviewing this material is that normalizing with the euclidean norm is preferred and it can be backed up mathematically.
